Towny/src/main/java/com/palmergames/bukkit/towny/permissions/TownyPerms.java

```diff
@@ -385,6 +391,12 @@ public static List<String> getTownRanks() {
 		return new ArrayList<String>(((MemorySection) perms.get("towns.ranks")).getKeys(false));
 	}
 
+	public static List<String> getTownRanks(Town town) {
+		if (!ranksWithTownLevelRequirementPresent())
+			return getTownRanks();
+		return getTownRanks().stream().filter(rank -> getRankTownLevelReq(rank) <= town.getLevelNumber()).collect(Collectors.toList());

```

Comment:
The level number can be cached here and in the nation method to avoid repeatedly calling the level calculate event